File: test/e2e/Apps/BasicDotNetIsolated/ActivityErrorHandling.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Concurrent;
using Microsoft.Azure.Functions.Worker;
using Microsoft.DurableTask;

namespace Microsoft.Azure.Durable.Tests.E2E;

public static class ActivityErrorHandling
{
    private static ConcurrentDictionary<string, int> globalRetryCount = new ConcurrentDictionary<string, int>();

    [Function(nameof(RethrowActivityException))]
    public static async Task<string> RethrowActivityException(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var output = await context.CallActivityAsync<string>(nameof(RaiseException), context.InstanceId);
        return output;
    }

    [Function(nameof(CatchActivityException))]
    public static async Task<string> CatchActivityException(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        try 
        {
            var output = await context.CallActivityAsync<string>(nameof(RaiseException), context.InstanceId);
            return output;
        }
        catch (TaskFailedException ex)
        {  
            return ex.Message;
        }
    }

    [Function(nameof(CatchActivityExceptionFailureDetails))]
    public static async Task<TaskFailureDetails?> CatchActivityExceptionFailureDetails(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        try 
        {
            await context.CallActivityAsync<string>(nameof(RaiseException), context.InstanceId);
            return null;
        }
        catch (TaskFailedException ex)
        {      
            return ex.FailureDetails;
        }
    }

    [Function(nameof(RetryActivityFunction))]
    public static async Task<string> RetryActivityFunction(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var options = TaskOptions.FromRetryPolicy(new RetryPolicy(
            maxNumberOfAttempts: 3,
            firstRetryInterval: TimeSpan.FromSeconds(3)));

        var output = await context.CallActivityAsync<string>(nameof(RaiseException), context.InstanceId, options: options);
        return output;
    }

    [Function(nameof(CustomRetryActivityFunction))]
    public static async Task<string> CustomRetryActivityFunction(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var options = TaskOptions.FromRetryHandler(retryContext => {
            if (retryContext.LastFailure.IsCausedBy<InvalidOperationException>() && 
                    retryContext.LastFailure.InnerFailure is not null && 
                    retryContext.LastFailure.InnerFailure.IsCausedBy<OverflowException>() && 
                    retryContext.LastAttemptNumber < 3) {
                return true;
            }
            return false;
        });

        string output = await context.CallActivityAsync<string>(nameof(RaiseComplexException), context.InstanceId, options: options);
        return output;
    }

    [Function(nameof(RaiseException))]
    public static string RaiseException([ActivityTrigger] string instanceId, FunctionContext executionContext)
    {
        if (globalRetryCount.AddOrUpdate(instanceId, 1, (key, oldValue) => oldValue + 1) == 1)
        {
            throw new InvalidOperationException("This activity failed");
        }
        else
        {
            return "Success";
        }
    }

    [Function(nameof(RaiseComplexException))]
    public static string RaiseComplexException([ActivityTrigger] string instanceId, FunctionContext executionContext)
    {
        if (globalRetryCount.AddOrUpdate(instanceId, 1, (key, oldValue) => oldValue + 1) == 1)
        {
            var exception = new InvalidOperationException("This activity failed\r\nMore information about the failure", innerException: new OverflowException("Inner exception message"));
            throw exception;
        }
        else
        {
            return "Success";
        }
    }
}

File: test/e2e/Apps/BasicJava/src/main/java/com/function/ActivityErrorHandling.java
package com.function;


import com.microsoft.azure.functions.annotation.*;
import com.function.exceptions.InvalidOperationException;
import com.function.exceptions.OverflowException;
import com.microsoft.azure.functions.*;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableActivityTrigger;
import com.microsoft.durabletask.azurefunctions.DurableOrchestrationTrigger;


/**
 * Azure Functions with Durable Task for activity error handling.
 */
public class ActivityErrorHandling {
    // Equivalent to static ConcurrentDictionary<string, int>
    private static final ConcurrentHashMap<String, AtomicInteger> globalRetryCount = new ConcurrentHashMap<>();

    /**
     * Orchestrator: RethrowActivityException
     */
    @FunctionName("RethrowActivityException")
    public String rethrowActivityException(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        // Calls RaiseException and lets any exception propagate
        return ctx.callActivity("RaiseException", ctx.getInstanceId(), String.class).await();
    }

    /**
     * Orchestrator: CatchActivityException
     */
    @FunctionName("CatchActivityException")
    public String catchActivityException(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        try {
            return ctx.callActivity("RaiseException", ctx.getInstanceId(), String.class).await();
        } catch (TaskFailedException ex) {
            return ex.getMessage();
        }
    }

    /**
     * Orchestrator: CatchActivityExceptionFailureDetails
     */
    @FunctionName("CatchActivityExceptionFailureDetails")
    public FailureDetails catchActivityExceptionFailureDetails(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        try {
            ctx.callActivity("RaiseException", ctx.getInstanceId(), String.class).await();
            return null;
        } catch (TaskFailedException ex) {
            return ex.getErrorDetails();
        }
    }

    /**
     * Orchestrator: RetryActivityFunction
     */
    @FunctionName("RetryActivityFunction")
    public String retryActivityFunction(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        RetryPolicy retryPolicy = new RetryPolicy(3, java.time.Duration.ofSeconds(3));
        TaskOptions options = new TaskOptions(retryPolicy);
        return ctx.callActivity("RaiseException", ctx.getInstanceId(), options, String.class).await();
    }

    /**
     * Orchestrator: CustomRetryActivityFunction
     */
    @FunctionName("CustomRetryActivityFunction")
    public String customRetryActivityFunction(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        TaskOptions options = new TaskOptions(retryContext -> {
            FailureDetails lastFailure = retryContext.getLastFailure();
            if (lastFailure != null
                    // BUG: https://github.com/microsoft/durabletask-java/issues/238
                    // BUG: https://github.com/microsoft/durabletask-java/issues/239
                    && lastFailure.getErrorType().equals("OverflowException")
                    // && lastFailure.isCausedBy(InvalidOperationException.class)
                    // && lastFailure.getInnerFailure() != null
                    // && lastFailure.getInnerFailure().isCausedBy("java.lang.OverflowException")
                    && retryContext.getLastAttemptNumber() < 3) {
                return true;
            }
            return false;
        });
        return ctx.callActivity("RaiseComplexException", ctx.getInstanceId(), options, String.class).await();
    }

    /**
     * Activity: RaiseException
     * @throws InvalidOperationException 
     */
    @FunctionName("RaiseException")
    public String raiseException(
            @DurableActivityTrigger(name = "instanceId") String instanceId,
            final ExecutionContext context) throws InvalidOperationException {
        AtomicInteger count = globalRetryCount.computeIfAbsent(instanceId, k -> new AtomicInteger(0));
        int current = count.incrementAndGet();
        if (current == 1) {
            throw new InvalidOperationException("This activity failed");
        } else {
            return "Success";
        }
    }

    /**
     * Activity: RaiseComplexException
     * @throws InvalidOperationException 
     */
    @FunctionName("RaiseComplexException")
    public String raiseComplexException(
            @DurableActivityTrigger(name = "instanceId") String instanceId,
            final ExecutionContext context) throws InvalidOperationException {
        AtomicInteger count = globalRetryCount.computeIfAbsent(instanceId, k -> new AtomicInteger(0));
        int current = count.incrementAndGet();
        if (current == 1) {
            OverflowException inner = new OverflowException("Inner exception message");
            InvalidOperationException ex = new InvalidOperationException(
                "This activity failed\r\nMore information about the failure", inner);
            throw ex;
        } else {
            return "Success";
        }
    }
}